### DISCUSSION TITLE
implementing tabbed-like window stacks, like pop shell

### DISCUSSION THREAD

**Discussion by sound-friend** [YaLTeR/niri#698] [Ideas]
![image](https://github.com/user-attachments/assets/26e2c8b7-a5d0-4862-af5c-a8c77b65f995)
something like this. my usual setup is one primary window and one for messaging, but some are seperate apps (like discord) and being able to stack these would be massively beneficial. let me know if you have any suggestions about this!

**Comment by YaLTeR** [comment]
I've heard this request from a few people already (smolsheep on matrix, sort of @sodiboo, @kchibisov). However, I never used tabbed windows much myself, and also certainly for niri we'll need to adjust the design somewhat. This is to say, I don't really know what the best way to do tabbed windows would be and what UX it should have (i.e. what the keyboard navigation should look like, since niri is more or less keyboard-centric).

Maybe some people can comment on how they envision using tabs and how they should work and we can discuss it?

**Reply by sodiboo** [reply]
I believe I've participated in discussion regarding tabbing in the past, but it's not actually something particularly high on my wishlist. They're a neat idea, but probably not quite the right fit for niri. I think most of what I would want out of tabbing can be better achieved by having better and more intuitive taskbars and similar such shell components, and some other niri functionalities like some sort of overview mode. That being said, I'm not against the concept of tabs. I'm probably not going to participate much in the initial design discussion about tabbing, but I may comment on whatever comes out of it.

**Reply by sound-friend** [reply]
i picture using it basically as regular scrolling but overlayed instead of next to each other. so super+right would go on the stack from left to right. this, as i said, would allow me to have a primary and secondary window section without having to manually swap windows.

**Reply by alicealysia** [reply]
> I've heard this request from a few people already (smolsheep on matrix, sort of @sodiboo, @kchibisov). However, I never used tabbed windows much myself, and also certainly for niri we'll need to adjust the design somewhat. This is to say, I don't really know what the best way to do tabbed windows would be and what UX it should have (i.e. what the keyboard navigation should look like, since niri is more or less keyboard-centric).
> 
> Maybe some people can comment on how they envision using tabs and how they should work and we can discuss it?

I tend to use a mix of several tools simultaneously while developing games and other complex applications, and being able to see the appropriate tools at the same time often demands being able to have one tabbed tile above another, or having multiple items on a row above another row. For example, having the scene heirachy on the left, an asset file manager below which can be tabbed to an error buffer. A scene view, drawing application, animation software, shader node editor, daw, text editor and 3d modelling software in the top pane, and then additional per object property controls on the right.

Having visibility of the appropriate tools at the appropriate aspect ratios in the right combinations is nightmarish with the current tabbed columns system, and it's grown from a nuisance to a serioua pain point, especially as so many are so dismissive of this need.

Generally my approach to this has been to use another window manager inside of niri that is built around non-scrolling tiling, but this introduces a lot of issues.

After giving it some thought, I believe the best approach is to introduce a "multi-row column display" type, which allows you to introduce multiple rows to a column, which then can next columns within themselves, including of course, tabbed columns. We could consider this to be akin to the existing column tab system, but with a different design. So, let's try cracking this settings and control design nightmare:

```
Binds {
 // converts column type into a multi row column
  Mod+y: "toggle-column-multirow-display" //y as in "y axis"
  Mod+y+= "add-row-to-multirow-display"
  Mod+y+- "remove-row-from-multirow-display" //merges current row with the row below, unless this is the bottom most row, in which case it merges with the above row
  //when moving up with Mod+up if you are at the topmost window of a column, and there is a row above, you will move to the leftmost column of that row, the same is true of moving downwards.
}
Layout {
  row-indicator {
   height 2 // this is a line between each row to denote that these are rows within a multirow-column rather than simply windows.
  gaps 2
  radius 0
  }
}
```

I think at least this should help keep it predictable from a user perspective.

**Reply by YaLTeR** [reply]
> For example, having the scene heirachy on the left, an asset file manager below which can be tabbed to an error buffer. A scene view, drawing application, animation software, shader node editor, daw, text editor and 3d modelling software in the top pane, and then additional per object property controls on the right.

Isn't this generally solved by the game engine app having a built-in panel tiling system?

I'm quite reluctant to add recursive tiling to niri, it's just too complex for its worth imo.

(If you're really determined, you could likely replicate arbitrary tiling layouts with scripting on top of floating windows, which always have coordinates in the IPC)

**Comment by nakibrayane** [comment]
I could see the utility of tabs in a traditional tiler, where tiling a lot of winodws in a small screen makes theme to small to use. but in a scrolling tiler  the windows don't get smaller by adding new windows.

Seeing your windows in a scrolling stack is in my opinion a better workflow then tabs. For example: if you have multiple terminals in tabs they will all be named the same thing (konsole - bash) and you would lose track of which tab has which terminal, looking at the content of the terminal is faster and way easier.

**Reply by sodiboo** [reply]
> if you have multiple terminals in tabs they will all be named the same thing (konsole - bash)

Terminals typically update window title based on the program running, or PWD when just in a prompt. SSH also adds a prefix showing the host. Terminal titles are generally distinct.

**Reply by kchibisov** [reply]
yeah, the issue with scrolling is that I have no clue about titles of my terminals since i don't see them. Urgency + title updates generally way things way more easy to navigate since you see everything at once, etc.

**Comment by kchibisov** [comment]
>However, I never used tabbed windows much myself, and also certainly for niri we'll need to adjust the design somewhat. This

tabbing +workspaces eliminates all the need for scrolling, since they both solve the issue of windows keeping the same size once you open something new and providing more space.

I'm not sure niri needs this, like those two things co-existing sound a bit strange, I'd say. Like you can surely have a notion of container like in sway and make tabs hard to use because of the amount of key strokes you have to press to get around and having scrolling on top, but it all just sounds like not a great UX.

But I wouldn't say that it's like entirely useless thing, it's just I don't see why you need both at the same time.

That also said, that niri will have to start drawing titlebars, etc.

**Comment by YaLTeR** [comment]
I actually have one awkward window management case which may be relevant.

I generally keep email, element, calendar, mastodon, discord as browser tabs. I like that they aren't *always visible*, and yet they are all easily accessible with a mouse click or a keyboard shortcut when I need them.

I kind of want to use Fractal or Tuba or other apps for these, but having them on a scrolling workspace is just awkward, because I don't need them *all at once* (and in case of email, I want it to be maximized), but scrolling back and forth to get to the right window is also very annoying. I also don't want to pick them on a taskbar window list type of thing.

Maybe WM tabs are a solution here? But then, maybe an overview + hot corner is the solution? When I used GNOME, I used to dump all of these apps into one workspace on top of each other, then use hot corner overview + click on window to super quickly switch to the right one, no tabs necessary.

**Reply by kchibisov** [reply]
Just having an easy way to switch to specific window(rofi's /apple spotlight switch to window thing) + overview will probably work way better than tabs for scrolling, since you can _scroll to_ nicely, where with tabs everything is usually just instant, etc, and navigation is basically switch workspace + bring window in tab group to front.

It's just sitting on 2 chairs is not a great way to maintain things and both tabbing and scrolling do a bit different UX when done in e.g. WM. Though, it's not like there're many tabbed WMs other than macOS.

**Reply by sound-friend** [reply]
what do you mean by wm tabs? as in the proposal i gave? nonetheless, i have a similar issue to you it sounds like. i think i personally use the scrolling funcionality different than others lol

**Reply by kchibisov** [reply]
Yeah, with scrolling you'd open just 2 maximized windows side by side(on your screenshot) and switch between, the same as tabs, you just don't see other window, so nothing really changes here...

**Reply by sound-friend** [reply]
my issue isnt necessarily the same as above, my setup is this:
![PXL_20240920_003844460 MP](https://github.com/user-attachments/assets/3c85cc98-d3cb-47e2-8864-2007e0c4c90b)
where i have a primary and secondary window. tabbing/overlaying apps on the right would simplify things i feel.
(note: this is gnome with paperwm)

**Reply by nakibrayane** [reply]
The way I wish this would be solved, is showing only the icons of the windows oppened in the current workspace in waybar, and have them arranged the same way they are in the column.

**Comment by rabidpug** [comment]
In my mind a way that tabs would complement nori's scrolling would be being able to toggle a single column of multiple windows to be tabbed, as that's the only instance where adding windows affects the space of other windows.
That way all the same key binds for interacting with windows within a column would work (so focus window up/down would scroll through the tabs, consume/expel would add or remove a window from the column)

**Reply by YaLTeR** [reply]
Interesting idea. The tabs could be vertical to signify this.

**Reply by sound-friend** [reply]
i feel like this would be a smart way to implement it. possibly by showing a stack at the bottom/top of previous and next ones in the "stack". this would compliment niri well i feel,

i also feel like this should be optional, and not enabled by default (?)

**Reply by rabidpug** [reply]
I like that idea, like the good old accordion.
And yea, in my mind I was thinking a keybind to toggle it on/off on a per-column basis, then a config that could define what the default should be for new multi-window columns (with the default being to maintain current behaviour).
Perhaps that config could be something like  `tabbed-column-when-windows-exceed #` so someone could configure it so a column of 2 windows by default will share the space, then 3+ would switch to be tabbed automatically.

**Reply by zetaPRIME** [reply]
ooh, yeah, this would be a really nice way to handle this; matches how I use things on i3 too (my i3 setup just tabs every workspace by default)
